lightning-graph-sync/src/processing.rs

```diff
@@ -0,0 +1,390 @@
+use core::ops::Deref;
+use std::io;
+use std::io::Read;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use lightning::chain;
+use lightning::ln::msgs;
+use lightning::ln::msgs::{ErrorAction, OptionalField, UnsignedChannelUpdate};
+use lightning::routing::network_graph;
+use lightning::routing::network_graph::ChannelMutationProperty;
+use lightning::util::ser::{BigSize, Readable};
+
+use crate::error::GraphSyncError;
+
+pub fn update_network_graph<CS: Deref>(
+    network_graph: &network_graph::NetworkGraph,
+    update_data: &[u8],
+    chain_source: &Option<CS>,
+) -> Result<(), GraphSyncError>
+where
+    CS::Target: chain::Access,
+{
+    let mut read_cursor = io::Cursor::new(update_data);
+    read_network_graph(&network_graph, &mut read_cursor, &chain_source)
+}
+
+pub(crate) fn read_network_graph<R: Read, CS: Deref>(
+    network_graph: &network_graph::NetworkGraph,
+    read_cursor: &mut R,
+    chain_source: &Option<CS>,
+) -> Result<(), GraphSyncError>
+where
+    CS::Target: chain::Access,
+{
+    let mut prefix = Vec::with_capacity(4);
+    prefix.resize(4, 0);
+    read_cursor.read_exact(&mut prefix)?;
+
+    let _deserialize_unsigned = match &prefix[..] {
+        &[76, 68, 75, 2] => true,
+        _ => {
+            return Err(GraphSyncError::ProcessingError(
+                "Prefix must equal 'LDK' and the byte 2".to_string(),
+            ));
+        }
+    };
+
+    // shadow variable for ownership release after length-restricted reads
+    let mut read_cursor = read_cursor;
+
+	let announcement_count: BigSize = Readable::read(read_cursor)?;
+	for _ in 0..announcement_count.0 {
+		let message_length: BigSize = Readable::read(read_cursor)?;
+		let mut restricted_cursor = read_cursor.take(message_length.0);
+		let channel_announcement: msgs::UnsignedChannelAnnouncement =
+			Readable::read(&mut restricted_cursor)?;
```

Comment:
Hmmm, we can cut the size here by a good chunk if we drop the two on-chain public keys. We only need the node public keys. I'm also curious if there's a way to avoid including the full node public keys for nodes that already have a channel (ie by providing an SCID + direction flag). I'm not sure if that'd be too slow on the decode side, though, need benchmarks.

Reply:
done